File: android/app/src/main/java/com/lichso/app/util/MoonPhaseCalculator.kt
package com.lichso.app.util

import kotlin.math.floor

/**
 * Pha Mặt Trăng
 */
object MoonPhaseCalculator {

    data class MoonPhase(val icon: String, val name: String, val age: Double)

    fun getMoonPhase(dd: Int, mm: Int, yy: Int): MoonPhase {
        val jd = LunarCalendarUtil.jdFromDate(dd, mm, yy)
        var k = ((jd - 2415021.076998695) / 29.530588853).toInt()
        var nmJd = LunarCalendarUtil.getNewMoonDay(k, LunarCalendarUtil.TZ)
        if (nmJd > jd) {
            k--
            nmJd = LunarCalendarUtil.getNewMoonDay(k, LunarCalendarUtil.TZ)
        }
        val moonAge = (jd - nmJd).toDouble()

        return when {
            moonAge < 1.5 -> MoonPhase("●", "Trăng mới", moonAge)
            moonAge < 7.4 -> MoonPhase("◑", "Trăng lưỡi liềm đầu", moonAge)
            moonAge < 8.4 -> MoonPhase("◑", "Bán nguyệt đầu", moonAge)
            moonAge < 14.4 -> MoonPhase("◕", "Trăng khuyết đầu", moonAge)
            moonAge < 15.8 -> MoonPhase("○", "Trăng tròn", moonAge)
            moonAge < 21.8 -> MoonPhase("◔", "Trăng khuyết cuối", moonAge)
            moonAge < 22.8 -> MoonPhase("◐", "Bán nguyệt cuối", moonAge)
            moonAge < 28.5 -> MoonPhase("◐", "Trăng lưỡi liềm cuối", moonAge)
            else -> MoonPhase("●", "Trăng mới", moonAge)
        }
    }
}

File: ios/LichSo/LichSo/Util/MoonPhaseCalculator.swift
import Foundation

// ═══════════════════════════════════════════
// Pha Mặt Trăng
// Port chính xác từ Android Kotlin
// KHÔNG ĐƯỢC THAY ĐỔI THUẬT TOÁN
// ═══════════════════════════════════════════

struct MoonPhase {
    let icon: String
    let name: String
    let age: Double
}

enum MoonPhaseCalculator {

    static func getMoonPhase(dd: Int, mm: Int, yy: Int) -> MoonPhase {
        let jd = LunarCalendarUtil.jdFromDate(dd: dd, mm: mm, yy: yy)
        var k = Int((Double(jd) - 2415021.076998695) / 29.530588853)
        var nmJd = LunarCalendarUtil.getNewMoonDay(k: k, timeZone: LunarCalendarUtil.TZ)
        if nmJd > jd {
            k -= 1
            nmJd = LunarCalendarUtil.getNewMoonDay(k: k, timeZone: LunarCalendarUtil.TZ)
        }
        let moonAge = Double(jd - nmJd)

        switch moonAge {
        case ..<1.5:   return MoonPhase(icon: "●", name: "Trăng mới", age: moonAge)
        case ..<7.4:   return MoonPhase(icon: "◑", name: "Trăng lưỡi liềm đầu", age: moonAge)
        case ..<8.4:   return MoonPhase(icon: "◑", name: "Bán nguyệt đầu", age: moonAge)
        case ..<14.4:  return MoonPhase(icon: "◕", name: "Trăng khuyết đầu", age: moonAge)
        case ..<15.8:  return MoonPhase(icon: "○", name: "Trăng tròn", age: moonAge)
        case ..<21.8:  return MoonPhase(icon: "◔", name: "Trăng khuyết cuối", age: moonAge)
        case ..<22.8:  return MoonPhase(icon: "◐", name: "Bán nguyệt cuối", age: moonAge)
        case ..<28.5:  return MoonPhase(icon: "◐", name: "Trăng lưỡi liềm cuối", age: moonAge)
        default:       return MoonPhase(icon: "●", name: "Trăng mới", age: moonAge)
        }
    }
}
